lib/mobility-core/src/Kernel/Storage/Esqueleto/DeletedEntity.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+module Kernel.Storage.Esqueleto.DeletedEntity
+  ( module Kernel.Storage.Esqueleto.DeletedEntity,
+    module Reexport,
+  )
+where
+
+import qualified Database.Esqueleto.Internal.Internal as Esq (SqlSelect)
+import Database.Persist.Postgresql as Esq (tableName)
+import Kernel.Prelude
+import qualified Kernel.Storage.Esqueleto as Esq
+import qualified Kernel.Storage.Esqueleto.DTypeBuilder as Esq
+import qualified Kernel.Storage.Esqueleto.Queries as Esq
+import Kernel.Storage.Tabular.DeletedEntity (DeletedEntityT)
+import Kernel.Types.DeletedEntity as Reexport
+import Kernel.Types.Id
+import Kernel.Utils.Common
+
+buildDeletedEntity :: forall t m. (MonadGuid m, MonadTime m, Esq.PersistEntity t, ToJSON t) => DeletedBy -> t -> m DeletedEntity
+buildDeletedEntity deletedBy entity = do
+  id <- Id <$> generateGUID
+  now <- getCurrentTime
+  let primaryId = Esq.keyFromRecordM <&> (\keyFromRecord -> encodeToText $ keyFromRecord entity)
+  pure
+    DeletedEntity
+      { id,
+        primaryId,
+        tableName = Esq.tableName entity,
+        rowData = encodeToText entity,
+        deletedBy = deletedBy,
+        deletedOn = now
+      }
+
+-- | Deleted entity will be automatically preserved into `deleted_entity` table
+deleteByIdP ::
+  forall t.
+  ( Typeable t,
+    Esq.TEntityKey t,
+    ToJSON t
```

Comment:
How it works? Our entities don't have ToJSON instances (At least the should not) directly, only `'Unsafe` ones have, but `'Unsafe` ones don't have TEntity instances

Reply:
I've added ToJSON instances for tabular types,  everywhere I use these logs. It works fine, I've tested.